### DISCUSSION TITLE
Detect location one of resources/data file

### DISCUSSION THREAD

**Discussion by emcek** [pyinstaller/pyinstaller#7598] [Help] [answered]
Hello,
Is there any way to call regular function inside `spec` file? I need add one of data file and its location depends on system configuration: it is inside current site-packages of virtual environment.

Something like:
``` python
def get_customtkinter_loc():
    ... 
    return ('venv311/Lib/site-packages/customtkinter', 'customtkinter')

a = Analysis(
    ['dcs_py.py'],
    pathex=[],
    binaries=[],
    datas=[get_customtkinter_loc(),
           ('dcspy/dcspy.ico', 'dcspy')],
    hiddenimports=[],
    hookspath=[],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    win_no_prefer_redirects=False,
    win_private_assemblies=False,
    cipher=block_cipher,
    noarchive=False,
)

**Comment by rokm** [comment, marked as answer]
> Is there any way to call regular function inside `spec` file?

Spec file is loaded and executed as python script, so yes.

**Reply by emcek** [reply]
in case of example above it does not work, or I miss something :/

**Reply by rokm** [reply]
Without a full example to work with and without knowing in what way it "does not work", my guess is as good as yours.

**Reply by rokm** [reply]
Although if you need to collect assets from `customtkinter` in environment's site-packages, you might want to use [`collect_data_files`](https://pyinstaller.org/en/stable/hooks.html#PyInstaller.utils.hooks.collect_data_files) function:

```python
from PyInstaller.utils.hooks import collect_data_files

datas = [('dcspy/dcspy.ico', 'dcspy')] + collect_data_files('customtkinter')

a = Analysis(
    ...
    datas=datas,
    ...
)
```

**Reply by emcek** [reply]
Yes, you are correct, my mistake:

```python
# -*- mode: python ; coding: utf-8 -*-
from pathlib import Path
from sys import prefix

packages = ['customtkinter', 'CTkMessagebox']
resources = ['dcspy.ico', 'dcspy_white.ico', 'config.yaml', 'falconded.ttf', 'dcspy.png', 'G13.png', 'G19.png', 'G510.png', 'G15v1.png', 'G15v2.png', 'license.txt']
files = [(f'dcspy/{r}', 'dcspy') for r in resources]
gui_packages = [(Path(prefix) / 'Lib' / 'site-packages' / f'{p}', p) for p in packages]

a = Analysis(
    ['dcs_py.py'],
    pathex=[],
    binaries=[],
    datas=files + gui_packages,
    ....
```
is working....

**Reply by bwoodsend** [reply]
> gui_packages = [(Path(prefix) / 'Lib' / 'site-packages' / f'{p}', p) for p in packages]

If you collect whole libraries as data files like that, you'll end up with all kinds of issues with duplicate libraries.

**Reply by emcek** [reply]
they are my dependencies from `requirements.txt`:
```python
packages = ['customtkinter', 'CTkMessagebox']
```
And I need add them manually since they using json file as theme descriptions. Pyinstaller ignore them and application complaining missing files by and therefore do not start.
https://customtkinter.tomschimansky.com/documentation/packaging
Do you have better way to do it?
![image](https://user-images.githubusercontent.com/475312/234783149-144b8474-06f0-42bf-8660-e799f7ed3bd0.png)

**Reply by bwoodsend** [reply]
Then you use `files += collect_data_files("customtkinter") + collect_data_files("CTkMessagebox")`.

**Reply by emcek** [reply]
@bwoodsend Perfect! Really thx for support!
